src/libmongoc/src/mongoc/mongoc-cursor-private.h

```diff
@@ -131,6 +132,17 @@ struct _mongoc_cursor_t {
    mongoc_read_prefs_t *read_prefs;
    mongoc_write_concern_t *write_concern;
 
+   /** If the cursor was created for an operation that might have overridden the
+    * user's read preferences' read mode, the actually-used read mode will be
+    * stored here. Otherwise, this is zero. This value should be restored to
+    * server_stream objects that are created from this cursor when the stream
+    * was created using the memoized 'server_id' of the cursor. */
+   mongoc_read_mode_t effective_read_mode;
+
+   /** Whether this cursor corresponds to an aggregate command that contains a
+    * writing-stage */
+   aggr_with_write_stage_flag is_aggr_with_write_stage;
```

Comment:
Both the name and the binary choice of enumerators available suggest this should be a `bool`.

Reply:
It may be a binary value, but at call sites it a plain `bool` is opaque and untyped, e.g.:

```c
retry_server_stream =
   mongoc_cluster_stream_for_reads (&client->cluster,
                                    parts->read_prefs,
                                    parts->assembled.session,
                                    NULL,
                                    NOT_AGGR_WITH_WRITE_STAGE,
                                    &ignored_error);
```

versus

```c
retry_server_stream =
   mongoc_cluster_stream_for_reads (&client->cluster,
                                    parts->read_prefs,
                                    parts->assembled.session,
                                    NULL,
                                    false,
                                    &ignored_error);
```

Reply:
All call sites use `NOT_AGGR_WITH_WRITE_STAGE`; `AGGR_WITH_WRITE_STAGE` is in fact completely unused in the code as proposed. Couldn't this parameter (and similar parameters in corresponding functions) be removed entirely? (This should alleviate concerns regarding the type/name of the data member on grounds of function call argument readability.)

Reply:
> `AGGR_WITH_WRITE_STAGE` is in fact completely unused in the code

This completely surprised me as "how could the code possibly work if no one ever passes this flag??" I dug through and realized that I was accidentally already implicitly relying on a `bool`-to-`enum` conversion in `mongoc-aggregate.c` to set the `mongoc_cursor_t::is_aggr_with_write_stage` attribute, which is then passed to `mongoc_cluster_stream_for_reads`, which is the single place where we _do_ actually pass `AGGR_WITH_WRITE_STAGE` after it is implicitly created from a `true` value. It's just a happy accident that the enum flag aligns with `true`/`false`, as inverting it would cause the logic to invert in the conversion. Oops. I was hoping that a compiler warning would have caught such a conversion, but it appears that no compiler implements such a warning ([and there doesn't seem to be much push to add such a warning](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=7654)). C enums hurt.

I'll fix the bool-to-enum conversion, but I suppose it can still be debated whether the binary `enum` is worth using.